### PR TITLE
Readiness endpoint

### PR DESCRIPTION
**Pull request by wibisono**
Updating snapshot status whenever Data Flusher updates snapshot.  Reusing health check end point with additional snapshot status with a ready field that will be true whenever snapshot size is more than reasonable minimum snapshot-size (configurable).

### CODE REVIEW

**Comment by bjpbakker** [comment]
Including this in the health check endpoint breaks the health check as a liveness indicator. This new snapshot-size check indicates readiness.

Liveness is used to determine whether the application is deployed properly, has the connectivity it needs, etc. If this fails, during deployment the container should be killed and deployment should stop (or retry a few times).

Readiness indicates when the newly deployed system should be switched to for public traffic - i.e. when the system is considered "up-to-date".

Additionally, I wonder if filesize is the right metric to use, perhaps the number of published objects is more important and more consistent.

File: src/main/resources/reference.conf [hunk 1]
```diff
@@ -75,3 +75,7 @@ locations.rsync = {
 unpublished-file-retain-period = 60m
 
 default.timeout = 10m
+
+# minimum snapshot size to indicate readiness of server
+minimum.snapshot.size = 100m
+minimum.snapshot.objects.count = 1000
```
[the remark this answers is not in the document]
Comment (lolepezy) [reply src/main/resources/reference.conf:80]: I am not sure which number should be here at all.
Is the counter reaches 1000, does it mean the server is ready?
In which sense it is "ready"?

**Comment by lolepezy** [comment]
What is the use case for the readiness? (I may have missed the discussion here)

File: src/main/scala/net/ripe/rpki/publicationserver/RRDPService.scala [hunk 2]
```diff
@@ -43,8 +43,14 @@ trait RRDPService extends RepositoryPath {
       get {
         complete(healthChecks.healthString)
       }
+    } ~ 
+    path("monitoring" / "readiness") {
+      get {
+        complete(healthChecks.readinessString)
```
[the remark this answers is not in the document]
Comment (wibisono) [reply src/main/scala/net/ripe/rpki/publicationserver/RRDPService.scala]: Updated to return service not ready when it's not complete. Also logged.

File: src/main/scala/net/ripe/rpki/publicationserver/repository/DataFlusher.scala [hunk 3]
```diff
@@ -292,16 +302,17 @@ class DataFlusher(conf: AppConfig)(implicit val system: ActorSystem)
...
   }
 
-  def withAtomicStream(targetDirectory: Path, secret: Bytes, filenameFromMac: Bytes => String, attrs: FileAttributes)(f : OutputStream => Unit) = {
+  def withAtomicStream(targetDirectory: Path, secret: Bytes, filenameFromMac: Bytes => String, attrs: FileAttributes)(f : OutputStream => Int) = {
     val tmpFile = Files.createTempFile(targetDirectory, "", ".xml", attrs)
     val tmpStream = new HashingSizedStream(secret, new FileOutputStream(tmpFile.toFile))
+    var objectsCount = 0
     try {
-      f(tmpStream)
+      objectsCount = f(tmpStream)
```
[the remark this answers is not in the document]
Comment (wibisono) [reply src/main/scala/net/ripe/rpki/publicationserver/repository/DataFlusher.scala]: Updated

File: src/main/scala/net/ripe/rpki/publicationserver/HealthChecks.scala [hunk 4]
```diff
@@ -18,14 +18,20 @@ class HealthChecks(val appConfig: AppConfig) {
...
   object HealthChecksJsonProtocol extends DefaultJsonProtocol {
     implicit val memoryFormat = jsonFormat3(Memory.apply)
     implicit val buildInformationFormat = jsonFormat4(BuildInformation.apply)
+    implicit val snapshotFormat = jsonFormat3(SnapshotStatus.apply)
     implicit val healthFormat = jsonFormat2(Health.apply)
   }
 
   lazy val objectStore = PgStore.get(appConfig.pgConfig)
 
+  var snapshotStatus = SnapshotStatus(false, 0, 0)
```
[the remark this answers is not in the document]
Comment (wibisono) [reply src/main/scala/net/ripe/rpki/publicationserver/HealthChecks.scala]: Using atomic integer now to keep object counts.

File: src/main/scala/net/ripe/rpki/publicationserver/HealthChecks.scala [hunk 5]
```diff
@@ -51,4 +62,8 @@ class HealthChecks(val appConfig: AppConfig) {
     val r = Runtime.getRuntime
     Memory(free = mb(r.freeMemory), total = mb(r.totalMemory), max = mb(r.maxMemory))
   }
+
+  def updateSnapshot(snapshotSize: Long, objectsCount: Int) = {
+    snapshotStatus = SnapshotStatus(snapshotSize > appConfig.minimumSnapshotSize && objectsCount > appConfig.minimumSnapshotObjectsCount, snapshotSize, objectsCount)
```
[the remark this answers is not in the document]
Comment (wibisono) [reply src/main/scala/net/ripe/rpki/publicationserver/HealthChecks.scala]: This is now refactored updating only count.

File: src/main/scala/net/ripe/rpki/publicationserver/HealthChecks.scala [hunk 6]
```diff
@@ -18,14 +18,20 @@ class HealthChecks(val appConfig: AppConfig) {
   case class Health(buildInformation: BuildInformation, databaseConnectivity: String)
   object Health
 
+  case class SnapshotStatus(ready: Boolean, size: Long, objectsCount: Int)
+  object SnapshotStatus
+
   object HealthChecksJsonProtocol extends DefaultJsonProtocol {
     implicit val memoryFormat = jsonFormat3(Memory.apply)
     implicit val buildInformationFormat = jsonFormat4(BuildInformation.apply)
+    implicit val snapshotFormat = jsonFormat3(SnapshotStatus.apply)
```
[the remark this answers is not in the document]
Comment (wibisono) [reply src/main/scala/net/ripe/rpki/publicationserver/HealthChecks.scala]: Refactored.

File: src/main/resources/reference.conf (hunk 1 above)
[the remark this answers is not in the document]
Comment (wibisono) [reply src/main/resources/reference.conf:80]: > What is the use case for the readiness? (I may have missed the discussion here)

When switching over we want the new target to not be empty. Not only that container is up and alive, but also ready to serve objects.

[the remark this answers is not in the document]
Comment (wibisono) [reply src/main/resources/reference.conf:80]: > I am not sure which number should be here at all.
> Is the counter reaches 1000, does it mean the server is ready?
> In which sense it is "ready"?

About the number, I don't know tbh what to put, maybe even 50K?

File: src/main/scala/net/ripe/rpki/publicationserver/repository/DataFlusher.scala [hunk 7]
```diff
@@ -18,7 +18,7 @@ import java.util.{Date, UUID}
 import scala.util.control.NonFatal
 
 
-class DataFlusher(conf: AppConfig)(implicit val system: ActorSystem)
+class DataFlusher(conf: AppConfig)(implicit val system: ActorSystem, implicit val healthChecks: HealthChecks)
```
[the remark this answers is not in the document]
Comment (wibisono) [reply src/main/scala/net/ripe/rpki/publicationserver/repository/DataFlusher.scala]: Removed the implicit.

**Pull request merged**